File: cb/dtos/controller.py
from __future__ import annotations
"""Controller DTO."""

import dataclasses
from typing import Any
from cb.dtos.base import BaseDTO


@dataclasses.dataclass
class ControllerDTO(BaseDTO):
    name: str = ""
    url: str = ""
    description: str = ""
    class_name: str = ""
    online: bool = True

    @classmethod
    def from_dict(cls, data: dict) -> "ControllerDTO":
        return cls(
            name=data.get("name", ""),
            url=data.get("url", ""),
            description=data.get("description", ""),
            class_name=data.get("_class", ""),
            online=not data.get("offline", False),
        )

File: src/core/dtos/controller.ts
/**
 * Controller DTO.
 * Mirrors legacy/cb/dtos/controller.py
 */

import { str, bool } from "./base.js";

/** Shape returned by the Operations Center controller list (`/cjoc/...` items). */
export interface ControllerDTO {
  name: string;
  url: string;
  description: string;
  className: string;
  online: boolean;
}

/** Maps a raw controller API entry to ControllerDTO. `_class` → `className`; `offline` is inverted to `online`. */
export function controllerFromDict(data: Record<string, unknown>): ControllerDTO {
  return {
    name: str(data["name"]),
    url: str(data["url"]),
    description: str(data["description"]),
    className: str(data["_class"]),
    online: !bool(data["offline"], false),
  };
}
